### ISSUE TITLE
Etcd3 -> etcd 2 rollback: etcd crashes

### ISSUE THREAD

**Issue by bprashanth** [kubernetes/kubernetes#32253]
Noticed fiddling with live cluster. etcd datadir and controller/apiserver logs: 
https://console.cloud.google.com/storage/browser/devnul/etcd-crash

Etcd logs are full of:
```
2016-09-08 02:01:00.533899 I | etcdmain: etcd Version: 2.2.1
2016-09-08 02:01:00.533949 I | etcdmain: Git SHA: 75f8282
2016-09-08 02:01:00.533953 I | etcdmain: Go Version: go1.5.1
2016-09-08 02:01:00.533957 I | etcdmain: Go OS/Arch: linux/amd64
2016-09-08 02:01:00.533964 I | etcdmain: setting maximum number of CPUs to 1, total number of available CPUs is 1
2016-09-08 02:01:00.534010 N | etcdmain: the server is already initialized as member before, starting as etcd member...
2016-09-08 02:01:00.534068 I | etcdmain: listening for peers on http://127.0.0.1:2380
2016-09-08 02:01:00.534088 I | etcdmain: listening for client requests on http://127.0.0.1:4001
2016-09-08 02:01:00.535059 W | snap: skipped unexpected non snapshot file db
2016-09-08 02:01:00.535071 I | etcdserver: name = default
2016-09-08 02:01:00.535075 I | etcdserver: data dir = /var/etcd/data
2016-09-08 02:01:00.535078 I | etcdserver: member dir = /var/etcd/data/member
2016-09-08 02:01:00.535082 I | etcdserver: heartbeat = 100ms
2016-09-08 02:01:00.535086 I | etcdserver: election = 1000ms
2016-09-08 02:01:00.535089 I | etcdserver: snapshot count = 10000
2016-09-08 02:01:00.535097 I | etcdserver: advertise client URLs = http://127.0.0.1:4001
2016-09-08 02:01:00.535127 W | wal: ignored file 0.tmp in wal
panic: runtime error: makeslice: len out of range

goroutine 1 [running]:
github.com/coreos/etcd/wal.(*decoder).decode(0xc820198630, 0xc820186178, 0x0, 0x0)
	/home/vagrant/gopath/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/wal/decoder.go:55 +0x142
github.com/coreos/etcd/wal.(*WAL).ReadAll(0xc820131d40, 0xc820194a60, 0x15, 0x20, 0x2, 0xce2a822cea30bfca, 0x15b0, 0x0, 0x0, 0x0, ...)
	/home/vagrant/gopath/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/wal/wal.go:238 +0x442
github.com/coreos/etcd/etcdserver.readWAL(0xc820194960, 0x19, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc820131d40, 0xc820191750, 0x5, ...)
	/home/vagrant/gopath/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/etcdserver/storage.go:84 +0x227
github.com/coreos/etcd/etcdserver.restartNode(0xc82013a0e0, 0x0, 0x7f74ef1a1028, 0xc8200b1400, 0x0, 0x0, 0x0, 0x0)
	/home/vagrant/gopath/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/etcdserver/raft.go:286 +0xff
github.com/coreos/etcd/etcdserver.NewServer(0xc82013a0e0, 0xc82013a100, 0x0, 0x0)
	/home/vagrant/gopath/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/etcdserver/server.go:288 +0x322d
github.com/coreos/etcd/etcdmain.startEtcd(0xc8200fa6c0, 0x0, 0x0, 0x0)
	/home/vagrant/gopath/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:287 +0x1a92
github.com/coreos/etcd/etcdmain.Main()
	/home/vagrant/gopath/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/etcdmain/etcd.go:117 +0x1e26
main.main()
	/home/vagrant/gopath/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/main.go:29 +0x14

goroutine 12 [chan receive]:
github.com/coreos/etcd/Godeps/_workspace/src/github.com/golang/glog.(*loggingT).flushDaemon(0x11b4240)
	/home/vagrant/gopath/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/github.com/golang/glog/glog.go:879 +0x67
created by github.com/coreos/etcd/Godeps/_workspace/src/github.com/golang/glog.init.1
	/home/vagrant/gopath/src/github.com/coreos/etcd/release/etcd/gopath/src/github.com/coreos/etcd/Godeps/_workspace/src/github.com/golang/glog/glog.go:410 +0x297

goroutine 32 [syscall]:
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x18
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x37
2016-09-08 02:01:02.244576 I | etcdmain: etcd Version: 2.2.1
2016-09-08 02:01:02.249642 I | etcdmain: Git SHA: 75f8282
2016-09-08 02:01:02.249658 I | etcdmain: Go Version: go1.5.1
2016-09-08 02:01:02.249662 I | etcdmain: Go OS/Arch: linux/amd64
2016-09-08 02:01:02.249670 I | etcdmain: setting maximum number of CPUs to 1, total number of available CPUs is 1
```

@kubernetes/sig-scalability @wojtek-t 
we probably at least want to triage for 1.4 so marking as p0.

**Comment by wojtek-t** [comment]
@hongchaodeng @xiang90 - seems like a problem with etcd itself.; can you please take a look?

**Comment by xiang90** [comment]
This is a out of date version of etcd. We fixed quite a few WAL related issues since 2.2. Probably we should upgrade it to 2.3.7 at least.

**Comment by wojtek-t** [comment]
Hmm this is strange. I think that we are already using 3.0.4 release now.

@bprashanth - where exactly does those logs come from?

**Comment by bprashanth** [comment]
From a 1.3.6 cluster

**Comment by bprashanth** [comment]
Maybe we need a cherrypick? I think 1.3.6 is still supposed to be the important 1.3 subrelease (meaning all the ones from 1.3-1.3.6 had issues). From the logs attached, it kept stack tracing, so if we don't fix it I guess we should release some PSA about how to recover a cluster in this state?

**Comment by wojtek-t** [comment]
@bprashanth - yeah, we probably should cherrypick 1.3, though it should already be fixed in 1.4 and master. So it's only a matter of 1.3.

**Comment by bprashanth** [comment]
SG, converting milestone 1.4 to 1.3

**Comment by lavalamp** [comment]
Wait, so this is an etcd3 -> etcd2 rollback issue?

**Comment by lavalamp** [comment]
Is there a version of etcd2 which does not have these issues? I don't think we should promote 1.3.7 to etcd3--you'd just have the same rollback issue between 1.3.6 and 1.3.7 if we did that.

**Comment by xiang90** [comment]
@lavalamp try 2.3.7. The rollback tool rolls back etcd data to version 2.3.7.

**Comment by lavalamp** [comment]
I'm thinking it's easier/better to just roll back to etcd 2.3.7 in 1.4. We can roll forward again in a point release. @wojtek-t @timothysc Thoughts?

**Comment by lavalamp** [comment]
Otherwise we have to make TWO releases. And for GKE (and presumably everyone else) that's a lot of extra upgrading.

**Comment by timothysc** [comment]
> I'm thinking it's easier/better to just roll back to etcd 2.3.7 in 1.4. 

Why?  This incident was against 1.3.6 using an older etcd.

> We can roll forward again in a point release. @wojtek-t @timothysc Thoughts?

I'd prefer not to change course at this time.

**Comment by lavalamp** [comment]
@timothysc if my understanding is correct, you cannot currently roll back from 1.4 to 1.3 because of this. We need to do *something* to make that possible.

**Comment by timothysc** [comment]
So cherry-picking etcd 2.3.7 for 1.3.(X) is not viable?  
That seems low cost, fixes 1.3.X, and allows us to stay the course. 

To be perfectly clear, this is more a gke constraint then a project constraint.  As we can/do version stack components independently.

**Comment by bprashanth** [comment]
In the past things that happen during rollback also happen when master reboots at the wrong moment. If we're not sure this is 100% triggered by rollback, reverting in 1.4 might re-introduce a bad race.

**Comment by lavalamp** [comment]
This has happened to at least three people here doing 1.4->1.3 rollbacks as part of the upgrade test script.

**Comment by lavalamp** [comment]
OK, I talked with @timothysc. Here is our etcd roadmap as I understand it:

1.4.0: ship with etcd 2.3.7
1.4.N: (N is after any urgent bug fix patch releases) bump to etcd 3.0.x
1.4.N+1: do the offline migration to start using the etcd 3 storage format & API.
1.5.0: do the migration to start storing proto natively (instead of JSON).

I will make a PR or find someone to make a PR to switch to etcd 2.3.7.

**Comment by timothysc** [comment]
@smarterclayton ^ we'll need to be mindful on rebase for 1.4.x here.

**Comment by lavalamp** [comment]
Built & pushed container:

```
dbsmith@danielsmith:~/code/k2/src/k8s.io/kubernetes/cluster/images/etcd$ make TAG=2.3.7 push
# Copy the content in this dir to the temp dir,
# without copying the subdirectories.
find ./ -maxdepth 1 -type f | xargs cp -t /tmp/tmp.OyfO19YfKd
make -C ../../../ WHAT=cluster/images/etcd/attachlease KUBE_STATIC_OVERRIDES="etcd";
make[1]: Entering directory `/usr/local/google/home/dbsmith/code/k2/src/k8s.io/kubernetes'
make[2]: Entering directory `/usr/local/google/home/dbsmith/code/k2/src/k8s.io/kubernetes'
make[2]: Leaving directory `/usr/local/google/home/dbsmith/code/k2/src/k8s.io/kubernetes'
make[2]: Entering directory `/usr/local/google/home/dbsmith/code/k2/src/k8s.io/kubernetes'
+++ [0908 14:05:10] Generating bindata:
    /usr/local/google/home/dbsmith/code/k2/src/k8s.io/kubernetes/test/e2e/framework/gobindata_util.go
+++ [0908 14:05:14] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [0908 14:05:16] Building go targets for linux/amd64:
    cmd/libs/go2idl/deepcopy-gen
+++ [0908 14:05:23] Generating bindata:
    /usr/local/google/home/dbsmith/code/k2/src/k8s.io/kubernetes/test/e2e/framework/gobindata_util.go
+++ [0908 14:05:26] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [0908 14:05:26] Building go targets for linux/amd64:
    cmd/libs/go2idl/conversion-gen
E0908 14:05:30.299567   14393 conversion.go:594] Warning: could not generate autoConvert functions for k8s.io/kubernetes/pkg/apis/apps/v1alpha1.PetSetSpec <-> k8s.io/kubernetes/pkg/apis/apps.PetSetSpec
E0908 14:05:30.660733   14393 conversion.go:594] Warning: could not generate autoConvert functions for k8s.io/kubernetes/pkg/apis/extensions/v1beta1.HorizontalPodAutoscalerSpec <-> k8s.io/kubernetes/pkg/apis/autoscaling.HorizontalPodAutoscalerSpec
E0908 14:05:30.671042   14393 conversion.go:594] Warning: could not generate autoConvert functions for k8s.io/kubernetes/pkg/apis/extensions/v1beta1.JobSpec <-> k8s.io/kubernetes/pkg/apis/batch.JobSpec
E0908 14:05:30.688036   14393 conversion.go:594] Warning: could not generate autoConvert functions for k8s.io/kubernetes/pkg/apis/extensions/v1beta1.RollingUpdateDeployment <-> k8s.io/kubernetes/pkg/apis/extensions.RollingUpdateDeployment
E0908 14:05:30.690839   14393 conversion.go:594] Warning: could not generate autoConvert functions for k8s.io/kubernetes/pkg/apis/extensions/v1beta1.ScaleStatus <-> k8s.io/kubernetes/pkg/apis/extensions.ScaleStatus
make[2]: Leaving directory `/usr/local/google/home/dbsmith/code/k2/src/k8s.io/kubernetes'
+++ [0908 14:05:31] Generating bindata:
    /usr/local/google/home/dbsmith/code/k2/src/k8s.io/kubernetes/test/e2e/framework/gobindata_util.go
+++ [0908 14:05:33] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [0908 14:05:33] Building go targets for linux/amd64:
    cluster/images/etcd/attachlease
make[1]: Leaving directory `/usr/local/google/home/dbsmith/code/k2/src/k8s.io/kubernetes'
cp "../../../_output/bin/attachlease" /tmp/tmp.OyfO19YfKd
# Do not compile if we should make an image for amd64, use the official etcd binaries instead
curl -sSL --retry 5 https://github.com/coreos/etcd/releases/download/v2.3.7/etcd-v2.3.7-linux-amd64.tar.gz | tar -xz -C /tmp/tmp.OyfO19YfKd --strip-components=1
# Replace BASEIMAGE with the real base image
cd /tmp/tmp.OyfO19YfKd && sed -i.bak 's|BASEIMAGE|busybox|g' Dockerfile
# And build the image
docker build -t gcr.io/google_containers/etcd-amd64:2.3.7 /tmp/tmp.OyfO19YfKd
Sending build context to Docker daemon  47.9 MB
Step 1 : FROM busybox
 ---> 8c2e06607696
Step 2 : MAINTAINER Dawn Chen <dawnchen@google.com>
 ---> Running in 2e5a3912287c
 ---> 0b0e43118f44
Removing intermediate container 2e5a3912287c
Step 3 : EXPOSE 2379 2380 4001 7001
 ---> Running in 4194fe392c73
 ---> c5b05849e64a
Removing intermediate container 4194fe392c73
Step 4 : COPY etcd etcdctl migrate-if-needed.sh attachlease /usr/local/bin/
 ---> 6ee3b2429a15
Removing intermediate container 16f586e6f6e5
Successfully built 6ee3b2429a15
gcloud docker push gcr.io/google_containers/etcd-amd64:2.3.7
The push refers to a repository [gcr.io/google_containers/etcd-amd64] (len: 1)
6ee3b2429a15: Pushed 
c5b05849e64a: Pushed 
0b0e43118f44: Pushed 
8c2e06607696: Pushed 
6ce2e90b0bc7: Pushed 
cf2616975b4a: Pushed 
2.3.7: digest: sha256:876a90f2d8711538c4cfb9f9a6da26634c449bcf27390d35fb63f3be4f6a6626 size: 9517
# Backward compatibility. TODO: deprecate this image tag
docker tag gcr.io/google_containers/etcd-amd64:2.3.7 gcr.io/google_containers/etcd:2.3.7
gcloud docker push gcr.io/google_containers/etcd:2.3.7
The push refers to a repository [gcr.io/google_containers/etcd] (len: 1)
6ee3b2429a15: Image already exists 
c5b05849e64a: Image already exists 
6ce2e90b0bc7: Image already exists 
2.3.7: digest: sha256:1fca894c78a2439f34722f4e62e6074d32010e970851e8a2c14740c926aad7d1 size: 9511
```

**Comment by lavalamp** [comment]
Change is simple. Hopefully we don't turn up any issues on etcd 2.3.7.

**Comment by lavalamp** [comment]
I'll also push the etcd 2.3.7 container for ARCH=arm, ARCH=arm64 and ARCH=ppc64le.

**Comment by lavalamp** [comment]
Hm, did I do something wrong? @luxas (who requested this in the PRs I'm cargo-culting)

```
$ make TAG=2.3.7 ARCH=arm push
# Copy the content in this dir to the temp dir,
# without copying the subdirectories.
find ./ -maxdepth 1 -type f | xargs cp -t /tmp/tmp.1XlBL7nThG
make -C ../../../ WHAT=cluster/images/etcd/attachlease KUBE_STATIC_OVERRIDES="etcd";
make[1]: Entering directory `/usr/local/google/home/dbsmith/code/k2/src/k8s.io/kubernetes'
make[2]: Entering directory `/usr/local/google/home/dbsmith/code/k2/src/k8s.io/kubernetes'
make[2]: Leaving directory `/usr/local/google/home/dbsmith/code/k2/src/k8s.io/kubernetes'
+++ [0908 15:32:54] Generating bindata:
    /usr/local/google/home/dbsmith/code/k2/src/k8s.io/kubernetes/test/e2e/framework/gobindata_util.go
+++ [0908 15:32:56] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
+++ [0908 15:32:57] Building go targets for linux/amd64:
    cluster/images/etcd/attachlease
make[1]: Leaving directory `/usr/local/google/home/dbsmith/code/k2/src/k8s.io/kubernetes'
cp "../../../_output/bin/attachlease" /tmp/tmp.1XlBL7nThG
# Download etcd in a golang container and cross-compile it statically
docker run -it -v /tmp/tmp.1XlBL7nThG:/etcdbin golang:1.6.3 /bin/bash -c \
                "git clone https://github.com/coreos/etcd /go/src/github.com/coreos/etcd \
                && cd /go/src/github.com/coreos/etcd \
                && git checkout v2.3.7 \
                && GOARM=6 GOARCH=arm ./build \
                && cp bin/arm/* /etcdbin"
Cloning into '/go/src/github.com/coreos/etcd'...
remote: Counting objects: 52271, done.
remote: Compressing objects: 100% (23/23), done.
remote: Total 52271 (delta 6), reused 0 (delta 0), pack-reused 52248
Receiving objects: 100% (52271/52271), 22.78 MiB | 789.00 KiB/s, done.
Resolving deltas: 100% (32803/32803), done.
Checking connectivity... done.
Note: checking out 'v2.3.7'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b new_branch_name

HEAD is now at fd17c91... version: bump to v2.3.7
cp: cannot stat 'bin/arm/*': No such file or directory
make: *** [build] Error 1
```

**Comment by smarterclayton** [comment]
@timothysc I hit this several times preparing the rebase and flipping between versions, glad that we tracked it down already.

**Comment by lavalamp** [comment]
We haven't actually tested that this resolves the issue yet--I'm going to get @grodrigues3 to test, since he has a cluster in this broken state.

**Comment by smarterclayton** [comment]
I was seeing forward failures as well from v2.3.0-284-gfbdb52d-> 3.0.6, where it failed to read the WAL.  Less impactful, but some kube users may hit it.

**Comment by xiang90** [comment]
@smarterclayton Have you tried 2.3.7 -> 3.0.6?

**Comment by lavalamp** [comment]
@smarterclayton good to know. Maybe we need to be hammering rollforward/back before we ship an etcd version.

@xiang90 how worried should we be about moving from 2.2.1 to 2.3.7?

**Comment by xiang90** [comment]
@lavalamp Yes. We should test it more extensively.

There should be no issues upgrading from 2.2.1 -> 2.3.7. Simply restarting etcd with new binary should work.

If there is an issue with upgrading from 2.3.7 -> 3.0.x, please report in etcd repo.

**Comment by smarterclayton** [comment]
My next step now that I saw this.  We were running older than that on
our previous version (but most customers run detached etcd anyway).

**Comment by lavalamp** [comment]
@xiang90 should 2.3.7 -> 2.2.1 be OK (rollback)?

**Comment by xiang90** [comment]
@lavalamp No. etcd does not support rollback.

**Comment by lavalamp** [comment]
@xiang90 Let's talk about this in realtime tomorrow if possible.

**Comment by lavalamp** [comment]
After chatting with @xiang90 I am not convinced we should change etcd versions at all. This requires us to do a backup before rolling forward, and I'm not sure if we can get that pulled together in time for a 1.4.0 release. I'm thinking we should stick with the current version and figure out a way to decouple etcd version from the release process. Otherwise we will be doing a bunch of 1.4.x releases just to manage etcd migrations.

**Comment by bprashanth** [comment]
so revert the etcd static pod file to what it is in 1.3.6 and figure out everything after 1.4, or are you proposing something else?

**Comment by bprashanth** [comment]
because it sounds like arbitrary etcd rollback is perilous, unless I misread between the lines

**Comment by wojtek-t** [comment]
I'm slightly lost in this whole discussion. Does it mean we can't go back from 2.3.7 to 2.2.1 (in case of problems)?
And if my understanding is correct, we should go to 2.3.7 first to enable upgrade to 3.0.x version.

**Comment by wojtek-t** [comment]
In other words, I think that we should go with 2.3.7 version in 1.4.0 release assuming it's possible to rollback to 2.2.1 from it (it's the same API and same storage format so it would be pretty bad if that's not possible...)

**Comment by lavalamp** [comment]
Right. It's *not* generally safe to rollback etcd versions. The supported method of rolling back is to start the old version from a backup.

To really do this correctly, we need to decouple etcd and apiserver versions. Because this means that you wouldn't be able to roll back from k8s 1.4.0 to 1.3.6 for a non-etcd problem.

**Comment by lavalamp** [comment]
So I think the only safe thing to do is lock in at etcd 2.2.1 and decouple this.

**Comment by bprashanth** [comment]
Reopening because we're now at 2.3.7

**Comment by liggitt** [comment]
the merged PR switched to 2.3.7... did we intend to go to 2.2.1 instead?

**Comment by lavalamp** [comment]
Yeah, I think we should go down to 2.2.1.
